### DISCUSSION TITLE
What's the point of level.BitStorage if I can not compact with it?

### DISCUSSION THREAD

**Discussion by maxsupermanhd** [Tnze/go-mc#196] [Q&A] [answered]
I expected data array to be initial data to be stored in compressed form but does not seem to be the case, do I need to write my own implementation of bit storage or will someone provide a method to do it?

**Comment by Tnze** [comment, marked as answer]
Why can't you write your init data with a loop?

**Reply by maxsupermanhd** [reply]
I just want to have my full uint32 array in clear form converted to compressed format, how do I do that? I don't want to go into the weeds of how it should be implemented, I think it is a library's job.

**Reply by Tnze** [reply]
uint32 array?

**Reply by Tnze** [reply]
[]uint64 > BitStorage: `NewBitStorage()`  
BitStorage > []uint64: `.Raw()`

**Reply by maxsupermanhd** [reply]
No it gets me compressed values, I will just do a bunch of .Set then

> []uint64 > BitStorage: `NewBitStorage()` BitStorage > []uint64: `.Raw()`

**Reply by Tnze** [reply]
Then call `Set()` in a for-loop. There is no shortcut. It needs calculation.

**Reply by maxsupermanhd** [reply]
That's what I am asking, why not have a method that accepts array to initialize in uncompressed form

**Reply by Tnze** [reply]
Just like you want to convert `[]int` into `list.List`. The memory layout are differents.

**Reply by Tnze** [reply]
In my opinion, `[]int` should not appear as an intermediate format in your code. Just read and write `BitStorage` directly.

**Reply by Tnze** [reply]
Unless you pursue some extreme performance, you need to use SIMD.

Otherwise, use `BitStorage`, lower memory footprint.
